slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala

```diff
@@ -1,16 +1,66 @@
 package slick.jdbc.hikaricp
 
-import java.sql.{Driver, Connection}
+import java.sql.{Connection, Driver}
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+
 import com.typesafe.config.Config
 import slick.SlickException
-import slick.jdbc.{JdbcDataSourceFactory, JdbcDataSource}
+import slick.jdbc.{JdbcDataSource, JdbcDataSourceFactory}
 import slick.util.ConfigExtensionMethods._
+import slick.util.Logging
 
 /** A JdbcDataSource for a HikariCP connection pool.
   * See `slick.jdbc.JdbcBackend#Database.forConfig` for documentation on the config parameters. */
-class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig) extends JdbcDataSource {
-  def createConnection(): Connection = ds.getConnection()
+class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig)
+  extends JdbcDataSource
+  with Logging {
+
+  private[this] val poolLock: ReentrantLock = new ReentrantLock()
+  private[this] val condition = poolLock.newCondition()
+  private[this] var count: Int = 0
+
+  def createConnection(): Connection = {
+    poolLock.lock()
+    try {
+      if (logger.isDebugEnabled) logger.debug("number of connections in use = "+count)
+      while (count == hconf.getMaximumPoolSize) {
+          condition.await()
+      }
+      val c = new DelegateConnection(ds.getConnection()) {
+        override def close(): Unit = {
+          closeConnection(conn)
+        }
+      }
+      count += 1
+      if (count == hconf.getMaximumPoolSize) {
+        logger.info("Maximum number of connections in use reached, pausing")
+        pause()
+      }
+      c
+    } finally {
+      poolLock.unlock()
+    }
+  }
+
+  private[this] def closeConnection(connection: Connection) = {
+    poolLock.lock()
+    try {
+      connection.close()
+    } finally {
+      count -= 1
+      if (count == hconf.getMaximumPoolSize -1) {
+        logger.info("Connection freed below maximum, resuming")
+        condition.signal()
+        resume()
+      }
+      poolLock.unlock()
+    }
+  }
+
+
   def close(): Unit = ds.close()
```

Comment:
I managed to write the following test. I'll see if I can add it to slicks test-kit module



class HikariCPJdbcDataSourceTest {

  @Test def testDatasourceShutdownDoesNotKeepThreadsLingering(): Unit = {
    val config: Config = ConfigFactory.load("application.conf").getConfig("h2mem1")
    val hikariCPJdbcDataSource: HikariCPJdbcDataSource = HikariCPJdbcDataSource.forConfig(config, null, "test", null)

    val max = hikariCPJdbcDataSource.hconf.getMaximumPoolSize
    val executor = new ThreadPoolExecutor(max, max, 1, TimeUnit.SECONDS, new LinkedBlockingQueue[Runnable]())
    implicit val executionContext = ExecutionContext.fromExecutor(executor)

    val latch = new CountDownLatch(max)

    //we schedule runnables to retrieve twice the amount of available connections
    //halve of them will be waiting for a connection to be released
    val eventualConnections: Seq[Future[Connection]] = 1 to max*2 map { _ =>
      Future {
        val c = hikariCPJdbcDataSource.createConnection()
        latch.countDown()
        c
      }
    }

    latch.await(10, TimeUnit.SECONDS)

    //halve of the threads have successfully retrieved a connection
    assertEquals(max, executor.getCompletedTaskCount)

    //halve of the threads should be blocked
    assertEquals(max, executor.getActiveCount)


    //shutdown the hikari datasource
    hikariCPJdbcDataSource.close()

    eventualConnections.foreach { f =>
      try {
        val conn: Connection = Await.result(f, Duration(30, TimeUnit.SECONDS))
        conn.close()
      } catch {
        case e: SQLException if e.getMessage == "Pool has been shutdown" => //ok
      }
    }

    //no more threads should be active
    assertEquals(0, executor.getActiveCount)
    //all threads have completed
    assertEquals(max*2, executor.getCompletedTaskCount)

  }

}